ios/FlankerNativeComponents/GameManager.swift

```diff
@@ -184,271 +124,257 @@ class GameManager {
+
+          if let image = URL(string: gameParameters.fixation), gameParameters.fixation.contains("https") {
+            delegate?.updateFixations(image: image, isStart: true, typeTime: .fixations)
+          } else {
+            delegate?.updateText(text: gameParameters.fixation, color: .black, font: Constants.bigFont, isStart: true, typeTime: .fixations)
+          }
+          timerSetText = Timer(timeInterval: gameParameters.fixationDuration / 1000, target: self, selector: #selector(setText), userInfo: nil, repeats: false)
+          RunLoop.main.add(timerSetText!, forMode: .common)
+      } else {
+        setText()
+      }
   }
 
   @objc func setText() {
-    guard let gameParameters = gameParameters else { return }
+      guard let gameParameters = gameParameters else { return }
+      guard countTest < gameParameters.trials.count else {
+          handleEndOfGame()
+          return
+      }
 
-    delegate?.setEnableButton(isEnable: true)
+      setEndTimeViewingImage(time: CACurrentMediaTime(), isStart: false, type: .fixations)
 
-    text = gameParameters.trials[countTest].stimulus.en
+      startTrialTimestamp = CACurrentMediaTime()
+      setEndTimeViewingImage(time: startTrialTimestamp!, isStart: true, type: .trial)
 
-    if let image = URL(string: text), text.contains("https") {
-      delegate?.updateFixations(image: image, isStart: true, typeTime: .trial)
-    } else {
-      delegate?.updateText(text: text, color: .black, font: Constants.bigFont, isStart: true, typeTime: .trial)
-    }
+      text = gameParameters.trials[countTest].stimulus.en
 
-    timeResponse = Timer(timeInterval: gameParameters.trialDuration / 1000, target: self, selector: #selector(self.timeResponseFailed), userInfo: nil, repeats: false)
-    RunLoop.main.add(timeResponse!, forMode: .common)
+      if let image = URL(string: text), text.contains("https") {
+          delegate?.updateFixations(image: image, isStart: true, typeTime: .trial)
+      } else {
+          delegate?.updateText(text: text, color: .black, font: Constants.bigFont, isStart: true, typeTime: .trial)
+      }
+
+      DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
+          self.delegate?.setEnableButton(isEnable: true)
+          self.timeResponse = Timer(timeInterval: gameParameters.trialDuration / 1000, target: self, selector: #selector(self.timeResponseFailed), userInfo: nil, repeats: false)
+          RunLoop.main.add(self.timeResponse!, forMode: .common)
+      }
   }
 
   @objc func timeResponseFailed() {
     guard let gameParameters = gameParameters else { return }
 
     delegate?.setEnableButton(isEnable: false)
+
+    endTrialTimestamp = CACurrentMediaTime()
+    setEndTimeViewingImage(time: endTrialTimestamp!, isStart: false, type: .trial)
+
+    startFeedbackTimestamp = CACurrentMediaTime()
+    setEndTimeViewingImage(time: startFeedbackTimestamp!, isStart: true, type: .feedback)
```

Comment:
The calls to `setEndTimeViewingImage` are duplicating the result of the previous line.

Does it make sense to keep the `setEndTimeViewingImage` method if you're setting it directly? It looks like you were trying to use it as a helper method.

Reply:
right, Since we're setting the timestamp variables directly before calling setEndTimeViewingImage, and the method also set the same variables, this results is redundant code.